tests/callbacks/test_quantization.py

```diff
@@ -139,3 +153,154 @@ def test_quantization_triggers(tmpdir, trigger_fn: Union[None, int, Callable], e
+        self._fake_quant_to_initial_state = {}
+        self._last_train_fake_quant_to_observer_enabled = {}
+
+    def _collect_observer_enabled(self) -> Dict[FakeQuantizeBase, Tensor]:
+        return {fake_quant: fake_quant.observer_enabled.clone() for fake_quant in self._fake_quants}
+
+    def _assert_fake_quant_initial_state(self):
+        # Check fake-quantization modules are restored to initial states.
+        for fake_quant, initial_state in self._fake_quant_to_initial_state.items():
+            state = fake_quant.state_dict()
+            for key, initial_value in initial_state.items():
+                assert torch.equal(state[key], initial_value)
+
+    def _assert_observer_enabled(self, expected_fake_quant_to_observer_enabled):
+        # Check observers belonging to fake-quantization modules are set properly.
+        fake_quant_to_observer_enabled = self._collect_observer_enabled()
+        for fake_quant in self._fake_quants:
+            observer_enabled = fake_quant_to_observer_enabled[fake_quant]
+            expected_observer_enabled = expected_fake_quant_to_observer_enabled[fake_quant]
+            if isinstance(expected_observer_enabled, bool):
+                # ``torch.quantization.FakeQuantize`` checks observer_enabled[0] == 1.
+                observer_enabled = observer_enabled[0] == 1
+                assert observer_enabled == expected_observer_enabled
+            else:
+                assert torch.equal(observer_enabled, expected_observer_enabled)
+
+    def _check_observer_state(self, should_disable):
+        if should_disable:
+            self._assert_observer_enabled(collections.defaultdict(lambda: False))
+        else:
+            self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+    def on_fit_start(self):
+        self._fake_quants = tuple(module for module in self.modules() if isinstance(module, FakeQuantizeBase))
+        self._fake_quant_to_initial_state = {
+            fake_quant: copy.deepcopy(fake_quant.state_dict()) for fake_quant in self._fake_quants
+        }
+        self._last_train_fake_quant_to_observer_enabled = self._collect_observer_enabled()
+
+    def training_step(self, batch, batch_idx):
+        self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+        result = super().training_step(batch, batch_idx)
+        if self._disable_observers_after_step == self.global_step:
+            self.apply(torch.quantization.disable_observer)
+
+        self._last_train_fake_quant_to_observer_enabled = self._collect_observer_enabled()
+
+        return result
+
+    def on_train_start(self):
+        self._last_train_fake_quant_to_observer_enabled = self._collect_observer_enabled()
+        self._check_observer_state("train" in self._disable_observers)
+
+    def on_train_end(self):
+        if "train" in self._disable_observers:
+            self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
```

Comment:
the hook impementations with assertions here are just mirroring what is implemented in the Callback. 
The complexity of this code is the same as the complexity of the callback, so an error here is equally likely as in the  callback implementations so in this sense the test is not doing much IMO. 

Can we not directly unit test the individual hook implementations?

Reply:
Will it be better to compare the model trained with `validate`, `test` and `predict` with the one trained with only training loop?

```python

@RunIf(quantization=True)
def test_quantization_val_test_predict(tmpdir):
    """Test quantization aware training not affected by valiation, testing and predicting"""
    seed_everything(42)
    num_features = 16
    dm = RegressDataModule(num_features=num_features)
    qmodel = RegressionModel()

    val_test_predict_qmodel = copy.deepcopy(qmodel)
    trainer = Trainer(
        callbacks=[QuantizationAwareTraining(quantize_on_fit_end=False)],
        default_root_dir=tmpdir,
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        limit_predict_batches=1,
        val_check_interval=1,
        num_sanity_val_steps=1,
        max_epochs=4
    )
    trainer.fit(val_test_predict_qmodel, datamodule=dm)
    trainer.validate(model=val_test_predict_qmodel, verbose=False)
    trainer.test(model=val_test_predict_qmodel, verbose=False)
    trainer.predict(
        model=val_test_predict_qmodel,
        dataloaders=[torch.utils.data.DataLoader(RandomDataset(num_features, 16))],
    )

    expected_qmodel = copy.deepcopy(qmodel)
    Trainer(
        callbacks=[QuantizationAwareTraining(quantize_on_fit_end=False)],
        default_root_dir=tmpdir,
        limit_train_batches=1,
        limit_val_batches=0,
        max_epochs=4
    ).fit(expected_qmodel, datamodule=dm)

    expected_state_dict = expected_qmodel.state_dict()
    for key, value in val_test_predict_qmodel.state_dict().items():
        expected_value = expected_state_dict[key]
        assert torch.allclose(value, expected_value)

```

Reply:
Yes, I think that could be good way. And if it is desired, one can additionally write some unit tests for the callback directly, like this:

```python
model = ...
trainer = ...
callback = QAT(...)

callback.on_validation_start(trainer, model) 
assert side_effect
assert side_effect
```

Reply:
The new tests will check that the default quantization aware training is not affected by validating, testing and predicting, and the callback hooks are checked separately.

Reply:
LGTM